### PR TITLE
IGAPP-605: Hooks for endpoints

### PR DESCRIPTION
**Pull request by klinzo**
This pull request belongs to an issue on our [bugtracker](https://issues.integreat-app.de/).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword **IGAPP**.

This PR allows creating hooks in the api-client as well as using them in the native and web-ts project. The `useLoadFromEndpoint` hook was moved from native to the api-client.

### CODE REVIEW

File: api-client/src/stringifyFormData.ts [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
 
   if (formData?.getParts) {
     for (const part of formData.getParts()) {
-      entries[part.fieldName] = part.string
+      entries[part.fieldName as string] = part.string
     }
   } else {
     for (const [key, value] of formData.entries()) {
-      entries[key] = value
+      entries[key as string] = value
```
Comment (Taggotty) [inline api-client/src/stringifyFormData.ts:11]: We probably should add tests for this. Especially as the types are a little bit hacky.

**Review by Taggotty — APPROVED** [review]

File: api-client/src/stringifyFormData.ts (hunk 1 above)
Comment (klinzo) [reply api-client/src/stringifyFormData.ts:11]: This is only used for showing a nice error, so I don't think this is critical. Maybe we could just use `JSON.stringify()` as well instead of this?

Comment (Taggotty) [reply api-client/src/stringifyFormData.ts:11]: Sure, it is not critical. Imo just JSON.stringify would be better, especially as we just print this message to the console right?

Comment (klinzo) [reply api-client/src/stringifyFormData.ts]: Replaced with JSON.stringify()

File: api-client/src/endpoints/hooks/useLoadFromEndpoint.ts [hunk 2]
```diff
@@ -1,24 +1,24 @@
...
-  setLoading: (arg0: boolean) => void
-) => {
+  setData: (data: T | null) => void,
+  setError: (error: Error | null) => void,
+  setLoading: (loading: boolean) => void
+): Promise<void> => {
   setLoading(true)
 
   try {
-    const apiUrl = await determineApiUrl()
```
Comment (SteffiStuffel) [inline api-client/src/endpoints/hooks/useLoadFromEndpoint.ts:14]: We used determineApiUrl for choosing test-cms or cms. Is this no longer needed? Or where gets it checked now?

File: package.json [hunk 3]
```diff
@@ -30,7 +30,8 @@
     "test-ts": "rm -rf */dist && tsc --build web-ts native"
   },
   "resolutions": {
-    "moment": "2.29.1"
+    "moment": "2.29.1",
+    "react": "16.13.1"
```
Comment (SteffiStuffel) [inline package.json:34]: Why is this necessary? Would really love it to be removed, if possible.

File: api-client/src/endpoints/hooks/useLoadFromEndpoint.ts (hunk 2 above)
Comment (SteffiStuffel) [reply api-client/src/endpoints/hooks/useLoadFromEndpoint.ts:14]: Oke I see it is in the Containers now. I think it would be a lot nicer if it would be central and only in one place. A container should not be in charge for determining the correct cms-url, should it? Is there a better place to put it?

**Review by SteffiStuffel — APPROVED** [review]

File: api-client/src/endpoints/hooks/useLoadFromEndpoint.ts (hunk 2 above)
Comment (klinzo) [reply api-client/src/endpoints/hooks/useLoadFromEndpoint.ts:14]: I moved `useLoadFromEndpoint` to the api-client. Therefore it is not possible to keep this functionality in the `useLoadFromEndpoint` hook. In theory it would be possible to create another wrapper around it in the native project though. I would not do this now already though.

File: package.json (hunk 3 above)
Comment (klinzo) [reply package.json:34]: It is necessary to define hooks in the api-client.

File: api-client/src/stringifyFormData.ts [hunk 4]
```diff
@@ -1,16 +0,0 @@
-// stringifies FormData from react-native or the browser variant of it
```
Comment (maxammann) [inline api-client/src/stringifyFormData.ts:1]: Are we sure this if fine to just use JSON.stringify?
